File: code/ex735.cpp
// Stack

class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        int n =  asteroids.size();
        
        stack<int> stk;
        for (int i = 0; i < n; i++) {
            while (!stk.empty() && stk.top() > 0 && asteroids[i] < 0 && stk.top() < -asteroids[i]) {
                stk.pop();
            }
            // When two asteroids have same size
            if (!stk.empty() && stk.top() > 0 && stk.top() == -asteroids[i]) {
                stk.pop();    continue;
            }
            // When stk.top() has the same sign with asteroids[i], push asteroids into the stack
            if (stk.empty() || stk.top() * asteroids[i] > 0 || stk.top() < 0)    stk.push(asteroids[i]);
        }
        
        // Get result
        vector<int> res(stk.size());
        for (int i = stk.size() - 1; i >= 0; i--) {
            res[i] = stk.top();    stk.pop();
        }
        
        return res;
    }
};

File: code/ex735.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        int n = asteroids.length;
        
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && stack.peek() > 0 && -asteroids[i] > stack.peek()) {
                stack.pop();
            }
            
            if (!stack.isEmpty() && stack.peek() > 0 && stack.peek() == -asteroids[i])    {stack.pop();    continue;}
            if (!stack.isEmpty() && stack.peek() * asteroids[i] < 0 && stack.peek() > 0)    continue;
            stack.push(asteroids[i]);
        }
        
        int[] res = new int[stack.size()];
        for (int i = stack.size() - 1; i >= 0; --i)
            res[i] = stack.pop();
        
        return res;
    }
}